From 1f7f87e2286262f8d98cbaba09a37764b789f605 Mon Sep 17 00:00:00 2001
From: Murala S Swamy <ssmurala@habana.ai>
Date: Thu, 19 Sep 2024 11:53:08 +0300
Subject: [PATCH] [FSW-8694] fs1: retrieve firmware security version

Added support to read security firmware version
from efuse device or from spi flash and check
minimum version from that data. For subsequent
module usage, store that into psoc register
and persistent dccm.

Change-Id: I35687a7ff3c90fdd4fde649c57d9c5d6a4cb408f
---

diff --git a/arch/fs1/arc/dev_security.c b/arch/fs1/arc/dev_security.c
index 27345e2..3044b71 100644
--- a/arch/fs1/arc/dev_security.c
+++ b/arch/fs1/arc/dev_security.c
@@ -12,6 +12,7 @@
 #include <fs1_efuse.h>
 #include <status.h>
 #include <dev_security.h>
+#include <security.h>
 
 /*
  * This key hash is calculated on the Public Key Properties -
@@ -85,6 +86,23 @@
 	boot_rom_global.sec_data.key_revoke = key_revoke;
 }
 
+void security_load_fw_version()
+{
+	uint8_t fw_version_sec = security_get_fw_version();
+
+	if (boot_rom_global.fake_fuse_en)
+		boot_rom_status_update(PROGRESS_VER_LOC_FAKE_FUSE);
+	else
+		boot_rom_status_update(PROGRESS_VER_LOC_ASIC_EFUSE);
+
+	/* store minimum allowed security version in PSOC */
+	regmap_writel(fw_version_sec,
+		IO0_H0_PSOC_BOOT_SECURITY(mmPSOC_SECURITY_MIN_FW_VERSION));
+
+	/* store minimum allowed security version in persistent DCCM */
+	boot_rom_global.sec_data.sec_ver = fw_version_sec;
+}
+
 int security_efuse_image_validation_en(void)
 {
 	/* check efuse for security validation enable */
diff --git a/arch/fs1/arc/dev_security.h b/arch/fs1/arc/dev_security.h
index 3b4cb98..c9067f6 100644
--- a/arch/fs1/arc/dev_security.h
+++ b/arch/fs1/arc/dev_security.h
@@ -28,6 +28,7 @@
 void security_load_revoke_bits(void);
 int security_efuse_image_validation_en(void);
 void boot_rom_eom_overrides(void);
+void security_load_fw_version(void);
 
 extern const uint8_t rom_key_hashes[NUM_KEYS_IN_ROM][SHA384_SUM_LEN];
 
diff --git a/arch/fs1/arc/progress_bits.h b/arch/fs1/arc/progress_bits.h
index 966f56a..ed09bcb 100644
--- a/arch/fs1/arc/progress_bits.h
+++ b/arch/fs1/arc/progress_bits.h
@@ -58,9 +58,9 @@
 	PROGRESS_KEY_LOC = 48,
 	PROGRESS_KEY_LOC_0 = 49,
 	PROGRESS_KEY_LOC_1 = 50,
-	PROGRESS_VER_LOC = 51,
-	PROGRESS_VER_LOC_0 = 52,
-	PROGRESS_VER_LOC_1 = 53,
+	PROGRESS_VER_LOC = 51, /* TODO: Remove & rearrange */
+	PROGRESS_VER_LOC_ASIC_EFUSE = 52,
+	PROGRESS_VER_LOC_FAKE_FUSE = 53,
 	PROGRESS_SLV_DIE_WAITING_ON_RECOVERY_IND = 54,
 	PROGRESS_SPI_MANIFEST_0_VERIFY_START = 55,
 	PROGRESS_SPI_MANIFEST_0_VERIFY_VERSION_FAIL = 56,
diff --git a/common/security.c b/common/security.c
index b4d00f9..4baa2f5 100644
--- a/common/security.c
+++ b/common/security.c
@@ -9,6 +9,10 @@
 #include <dev_security.h>
 #include <string.h>
 #include <security.h>
+#include <mem_io.h>
+#include <efuse.h>
+#include <limits.h>
+#include <fs1_efuse.h>
 
 int security_check_valid_fw_version(uint32_t fw_version)
 {
@@ -61,6 +65,7 @@
 
 	security_load_sec_efuse_dword();
 	security_load_revoke_bits();
+	security_load_fw_version();
 }
 
 int security_rom_get_key_hash_by_idx(uint32_t *hash, uint8_t key_index)
@@ -71,3 +76,43 @@
 
 	return 0;
 }
+
+/*
+ * minimal allowed security version is defined by a 128bit map, to find this
+ * we scan from end to beginning of map looking for the first set bit. once
+ * it's found this indicated the minimal version allowed to run on the system
+ */
+static uint8_t security_find_min_fw_version(uint32_t *fw_version)
+{
+	int8_t row, column;
+	uint8_t fw_version_sec;
+	/* init max version */
+	fw_version_sec = SECURITY_MAX_FW_SEC_VERSION;
+
+	/* iterate from end to start, till a set bit is found */
+	for (row = SECURITY_FW_VER_LAST_DWORD; row >= 0; row--) {
+		for (column = SECURITY_FW_VER_LAST_BIT; column >= 0; column--) {
+			/* find highest bit set in valid FW */
+			if (fw_version[row] & BIT(column))
+				return fw_version_sec;
+			fw_version_sec--;
+		}
+	}
+
+	return fw_version_sec;
+}
+
+uint8_t security_get_fw_version(void)
+{
+	uint8_t i;
+	uint32_t fw_version[IO_EFUSE_SVN_FW_DWORD_COUNT];
+	uint8_t fw_version_sec;
+
+	/* read min security version bit map from flash */
+	for (i = 0; i < IO_EFUSE_SVN_FW_DWORD_COUNT; i++)
+		boot_rom_fuse_read(IO_EFUSE_SVN_FW_OFFSET + i, &fw_version[i]);
+
+	fw_version_sec = security_find_min_fw_version(fw_version);
+
+	return fw_version_sec;
+}
diff --git a/common/security.h b/common/security.h
index 1d03c35..527f3f8 100644
--- a/common/security.h
+++ b/common/security.h
@@ -8,11 +8,16 @@
 
 #define SECURITY_FW_VERSION_SECURE		1
 #define SECURITY_FW_VERSION_NOT_SECURE		-1
-#define SECURITY_MAX_FW_SEC_VERSION		128
+
+#define SECURITY_DWORD_BIT_SIZE			(sizeof(uint32_t) * CHAR_BIT)
+#define SECURITY_FW_VER_LAST_BIT		(SECURITY_DWORD_BIT_SIZE - 1)
+#define SECURITY_FW_VER_LAST_DWORD		(IO_EFUSE_SVN_FW_DWORD_COUNT - 1)
+#define SECURITY_MAX_FW_SEC_VERSION		(IO_EFUSE_SVN_FW_DWORD_COUNT * SECURITY_DWORD_BIT_SIZE)
 
 void boot_rom_security_init(void);
 int security_check_key_revoke_status(uint32_t key_index);
 int security_rom_get_key_hash_by_idx(uint32_t *hash, uint8_t key_index);
 int security_check_valid_fw_version(uint32_t fw_version);
+uint8_t security_get_fw_version(void);
 
 #endif /* _SECURITY_H_ */
